### DISCUSSION TITLE
How to Retrieve Real-Time Output from UserProxyAgent.

### DISCUSSION THREAD

**Discussion by naphatkps** [microsoft/autogen#3571] [Q&A]
Hi, presently I'm involved with the development of an agentic web application using AutoGen. I've come to an issue that prevents me from getting the output of the `UserProxyAgent` right away.

Presently, I use `user_proxy.initiate_chat()`, which works through the terminal and the log is only accessible when the session is over. I want to dynamically get that output and show the same to the user real-time in the web application.

What would be the best way to do that? Any suggestions or examples would be very welcome!

Thanks in advance!

**Comment by KeepALifeUS** [comment]
For real-time output in web applications, here's a pattern using shared state:

## State-Based Streaming

```python
import asyncio
from queue import Queue

# Shared state for real-time output
output_queue = Queue()

class StreamingObserver:
    def __init__(self, queue):
        self.queue = queue
    
    def on_message(self, message):
        # Push each message to queue immediately
        self.queue.put({
            "type": "message",
            "sender": message.get("sender"),
            "content": message.get("content"),
            "timestamp": time.time()
        })

# Web endpoint that streams from queue
async def stream_output():
    while True:
        if not output_queue.empty():
            yield output_queue.get()
        await asyncio.sleep(0.1)
```

## Integration with AutoGen

```python
from autogen import AssistantAgent, UserProxyAgent

def custom_print_func(message):
    # Intercept messages and push to queue
    output_queue.put(message)
    print(message)  # Optional: still print to console

# Override the print function used by agents
assistant = AssistantAgent(
    name="assistant",
    human_input_mode="NEVER"
)
assistant._print_received_message = custom_print_func
```

## Why This Works

1. **No blocking** — Messages are queued immediately
2. **Real-time** — Frontend polls/SSE from queue
3. **Framework-agnostic** — Works with any web framework
4. **Full history** — Queue preserves order

The key insight: intercept messages at the print/logging layer, push to a shared queue, and stream from there.

More on state-based patterns: https://github.com/KeepALifeUS/autonomous-agents
